readme.md
# Angular Validation (Directive / Service)
`Version: 1.5.28`

## Project in Life Support
#### still accepting PR for any bug fix
**Also note that only NPM will updated with new releases when PR get merged**

This project is now in Life Support since most of us already moved to newer version of Angular. However I do want to point out that if you still use the lib and find a Bug, I certainly still welcome PR (Pull Request) to address bug fixes. So I'm not totally gone but I won't personally invest more time in the lib. Also note that the lib will not be rewritten to support Angular 2+

On a totally different note, I'm still very active in the Angular 4+ world (even Aurelia world) and you might be interested in looking at some of my other libraries.
- [Angular-Slickgrid](https://github.com/ghiscoding/Angular-Slickgrid)
- [Angular Markdown Preview Editor](https://github.com/ghiscoding/angular-markdown-editor)

In the Aurelia world
- [Aurelia-Slickgrid](https://github.com/ghiscoding/aurelia-slickgrid)
- [Aurelia Bootstrap Plugin](https://github.com/ghiscoding/Aurelia-Bootstrap-Plugins)

---

### Forms Validation with Angular made easy!
##### (Concept comes from the amazing Laravel)

Form validation after user stop typing (debounce default of 1sec). Angular-Validation is an angular directive/service with locales (languages) with a very simple approach of defining your `validation=""` directly within your element to validate (input, textarea, etc) and...that's it!!! The directive/service will take care of the rest!

The base concept is not new, it comes from the easy form input validation approach of Laravel Framework as well as PHP Gump Validation. They are both PHP frameworks and use a very simple approach, so why not re-use the same concept over Angular as well? Well it's now made available with a few more extras.

For a smoother user experience, I also added validation on inactivity (timer/debounce). So validation will not bother the user while he is still typing... though as soon as the user pauses for a certain amount of time, then validation comes into play. It's worth knowing that this inactivity timer is only available while typing, if user focuses away from his input (onBlur) it will then validate instantly.

Supporting AngularJS 1.3.x-1.5.x branch *(current code should work with 1.2.x just the same, but is no more verified)*

Now support <b>Service</b> using the same functionalities as the <b>Directive</b>.
Huge rewrite to have a better code separation and also adding support to Service functionalities. Specifically the `validation-rules` was separated to add rules without affecting the core while `validation-common` is for shared functions (shared by Directive/Service).

[Validation Summary](https://github.com/ghiscoding/angular-validation/wiki/Validation-Summary) was also recently added to easily show all validation errors that are still active on the form and you can also use 2 ways of dealing with the [Submit and Validation](https://github.com/ghiscoding/angular-validation/wiki/Form-Submit-and-Validation) button.

For more reasons to use it, see the answered question of: [Why Use It?](#whyuseit)

If you like the Angular-Validation project and you use it, please click on the :star: and add it as a favorite. The more star ratings there is, the more chances it could be found by other users inside the popular trend section. That is the only support I ask you... thanks and enjoy it ;)

<a name="plunker"></a>
## Live Demo
[Plunker](http://plnkr.co/jADq7H)

## Tested with Protractor
Angular-validation now has a full set of **End-to-End tests** with **Protractor**, there is over 1800+ assertions, it starts by testing the original live demo page and then goes on with a complete test suite of All Validators in both the Angular-Validation Directive and Service.

<a name="whyuseit"></a>
Why use angular-validation?
-----
Angular-validation was develop with simplicity and DRY (Don't Repeat Yourself) concept in mind.
You can transform this:
```html
<input type="text" name="username" ng-model="user.username" ng-minlength="3" ng-maxlength="8" required />
<div ng-show="form.$submitted || form.user.$touched">
  <span ng-show="userForm.username.$error.minlength" class="help-block">Username is too short.</p>
  <span ng-show="userForm.username.$error.maxlength" class="help-block">Username is too long.</p>
</div>
<input type="text" name="firstname" ng-model="user.firstname" ng-minlength="3" ng-maxlength="50" required />
<div ng-show="form.$submitted || form.user.$touched">
  <span ng-show="userForm.firstname.$error.minlength" class="help-block">Firstname is too short.</p>
  <span ng-show="userForm.firstname.$error.maxlength" class="help-block">Firstname is too long.</p>
</div>
<input type="text" name="lastname" ng-model="user.lastname" ng-minlength="2" ng-maxlength="50" required />
<div ng-show="form.$submitted || form.user.$touched">
  <span ng-show="userForm.lastname.$error.minlength" class="help-block">Lastname is too short.</p>
  <span ng-show="userForm.lastname.$error.maxlength" class="help-block">Lastname is too long.</p>
</div>
```
into the following (errors will automatically be displayed in your chosen locale translation):
```html
<input type="text" name="username" ng-model="user.username" validation="min_len:3|max_len:8|required"  />
<input type="text" name="firstname" ng-model="user.firstname" validation="alpha_dash|min_len:3|max_len:50|required"  />
<input type="text" name="lastname" ng-model="user.lastname" validation="alpha_dash|min_len:2|max_len:50|required"  />
```
The Angular-Validation will create, by itself, the necessary error message. Now imagine your form having 10 inputs, using the documented Angular way will end up being 30 lines of code, while on the other hand `Angular-Validation` will stay with 10 lines of code, no more... so what are you waiting for? Use Angular-Validation!!!  Don't forget to add it to your favorite, click on the :star: on top :)

Let's not forget the [Validation Summary](https://github.com/ghiscoding/angular-validation/wiki/Validation-Summary) which is also a great and useful way of displaying your errors to the user.

Another awesome feature recently added is the [Wiki - Remote Validation (AJAX)](https://github.com/ghiscoding/angular-validation/wiki/Remote-Validation-(AJAX)) which is useful for backend server validation.

Again another recently added feature, external 3rd party addon validation (like ngTagsInput or Angular Multi-Select), take a look at the [Wiki - 3rd party addon validation](https://github.com/ghiscoding/angular-validation/wiki/3rd-Party-Addons)


## Angular-Validation Wiki
All the documentation has been moved to the Wiki section, see the [github wiki](https://github.com/ghiscoding/angular-validation/wiki) for more explanation. If you just started with the library, then the most important page to read would be the [HOWTO - Step by Step](https://github.com/ghiscoding/angular-validation/wiki/HOWTO---Step-by-Step)

**Wiki Contents**
* [Angular-Validation Wiki](https://github.com/ghiscoding/angular-validation/wiki)
* Installation
    * [HOWTO - Step by Step](https://github.com/ghiscoding/angular-validation/wiki/HOWTO---Step-by-Step)
    * [Bower/NPM/NuGet Packages](https://github.com/ghiscoding/angular-validation/wiki/Download-and-Install-it)
    * [Locales (languages)](https://github.com/ghiscoding/angular-validation/wiki/Locales-(languages))
* Code Samples
    * [3rd Party Addon Validation](https://github.com/ghiscoding/angular-validation/wiki/3rd-Party-Addons)
    * [Directive Examples](https://github.com/ghiscoding/angular-validation/wiki/Working-Directive-Examples)
    * [Service Examples](https://github.com/ghiscoding/angular-validation/wiki/Working-Service-Examples)
    * [Bootstrap Decorator (has-error)](https://github.com/ghiscoding/angular-validation/wiki/Bootstrap-Decorator-(has-error))
* Functionalities
    * [Alternate Text on Validators](https://github.com/ghiscoding/angular-validation/wiki/Alternate-Text-on-Validators)
    * [DisplayErrorTo](https://github.com/ghiscoding/angular-validation/wiki/Bootstrap-Input-Groups-Wrapping)
    * [Isolated Scope](https://github.com/ghiscoding/angular-validation/wiki/Isolated-Scope)
    * [PreValidate Form (on load)](https://github.com/ghiscoding/angular-validation/wiki/PreValidate-Form-(on-page-load))
    * [Reset Form](https://github.com/ghiscoding/angular-validation/wiki/Reset-Form)
    * [Revalite an input triggered by another Input](https://github.com/ghiscoding/angular-validation/wiki/Revalidate-Input)
    * [Submit and Validation](https://github.com/ghiscoding/angular-validation/wiki/Form-Submit-and-Validation)
    * [Validation Callback](https://github.com/ghiscoding/angular-validation/wiki/Validation-Callback)
    * [Validator Remove](https://github.com/ghiscoding/angular-validation/wiki/Remove-Validator-from-Element)
    * [Validation Summary](https://github.com/ghiscoding/angular-validation/wiki/Validation-Summary)
* Custom Validations
    * [Custom Validation (JS)](https://github.com/ghiscoding/angular-validation/wiki/Custom-Validation-functions)
    * [Remote Validation (AJAX)](https://github.com/ghiscoding/angular-validation/wiki/Remote-Validation-(AJAX))
* Properties & Options
    * [Attributes (all options)](https://github.com/ghiscoding/angular-validation/wiki/Inputs-(local-options))
    * [Global Options](https://github.com/ghiscoding/angular-validation/wiki/Global-Options)
    * [ControllerAs Syntax](https://github.com/ghiscoding/angular-validation/wiki/ControllerAs-Syntax)
* Validators
    * [Available Validator Rules](https://github.com/ghiscoding/angular-validation/wiki/Available-Validators-(rules))
    * [Custom Regular Expression Pattern](https://github.com/ghiscoding/angular-validation/wiki/Regular-Expression-Pattern)
* Misc
    * [Changelog](https://github.com/ghiscoding/angular-validation/wiki/CHANGELOG)
    * [License](https://github.com/ghiscoding/angular-validation/wiki/License)

<a name="install"></a>
Download and Install it
-----
Install with **Bower**

```javascript
// bower install with
bower install angular-validation-ghiscoding
```
Install with **NPM**

```javascript
// NPM install with
npm install angular-validation-ghiscoding
```
Install with **NuGet** (see the [NuGet Package Here](http://www.nuget.org/packages/Angular-Validation-Ghiscoding))
```javascript
PM> Install-Package Angular-Validation-Ghiscoding
```
When used with IIS, you will need to map the JSON type
```html
<staticContent>
    <mimeMap fileExtension=".json" mimeType="application/json" />
</staticContent>
```

### License
[MIT License](http://www.opensource.org/licenses/mit-license.php)

### Available Validator Rules
All validators are written as `snake_case` but it's up to the user's taste and could also be used as `camelCase`. So for example `alpha_dash_spaces` and `alphaDashSpaces` are both equivalent.

##### NOTE: on an `input type="number"`, the `+` sign is an invalid character (browser restriction) even if you are using a `signed` validator. If you really wish to use the `+`, then change your input to a `type="text"`.

* `accepted` The field under validation must be `yes`, `on`, `1`, or `true`. Useful for validating "Terms of Service" acceptance.
* `alpha` Only alpha characters (including latin) are present (a-z, A-Z)
* `alpha_spaces` Only alpha characters (including latin) and spaces are present (a-z, A-Z)
* `alpha_num` Only alpha-numeric characters (including latin) are present (a-z, A-Z, 0-9)
* `alpha_num_spaces` Only alpha-numeric characters (with latin & spaces) are present (a-z, A-Z, 0-9)
* `alpha_dash` Only alpha-numeric characters + dashes, underscores are present (a-z, A-Z, 0-9, _-)
* `alpha_dash_spaces` Alpha-numeric chars + dashes, underscores and spaces (a-z, A-Z, 0-9, _-)
* `between:min,max` Will auto-detect value type then use proper validator.
  * Type Number uses `between_num`, String use `between_len`.
* `between_date_iso:d1,d2` Alias of `between_date_iso`.
* `between_date_euro:d1,d2` Alias of `date_euro_between`.
* `between_date_euro_long:d1,d2` Alias of `date_euro_long_between`.
* `between_date_euro_short:d1,d2` *DEPRECATED* does not support leap year, preferable to use `date_euro_between` or make a PR to fix it..
* `between_date_us:d1,d2` Alias of `date_us_between`.
* `between_date_us_long:d1,d2` Alias of `date_us_long_between`.
* `between_date_us_short:d1,d2` *DEPRECATED* does not support leap year, preferable to use `date_us` or make a PR to fix it..
* `between_len:min,max` Ensures the length of a string is between a min,max length.
* `between_num:min,max` Ensures the numeric value (int or float) is between a min,max number.
* `boolean` Ensures the value is `true` or `false` (`0` or `1` is also valid).
* `compare` Alias of `match`
* `credit_card` Valid credit card number (AMEX, VISA, Mastercard, Diner's Club, Discover, JCB)
* `date_iso` Ensure date follows the ISO format (yyyy-mm-dd)
* `date_iso_between:d1,d2` Ensure date follows the ISO format and is between (d1) &amp; (d2)
* `date_iso_max:d` Date must follow ISO format and is lower or equal than date (d)
* `date_iso_min:d` Date must follow ISO format and is higher or equal than date (d)
* `date_euro` Date must follow the European short or long format (dd-mm-yyyy) or (dd/mm/yyyy)
* `date_euro_long` Date must follow the European long format (dd-mm-yyyy) or (dd/mm/yyyy)
* `date_euro_between:d1,d2` Date must follow European short or long format and is between (d1) &amp; (d2)
* `date_euro_long_between:d1,d2` Date must follow European long format and is between (d1) &amp; (d2)
* `date_euro_max:d` Date must follow European short or long format and is lower or equal than date (d)
* `date_euro_long_max:d` Date must follow European long format and is lower or equal than date (d)
* `date_euro_min:d` Date must follow European short or long format and is higher or equal than date (d)
* `date_euro_long_min:d` Date must follow European long format and is higher or equal than date (d)
* `date_euro_short` *DEPRECATED* does not support leap year, preferable to use `date_euro` or make a PR to fix it..
* `date_euro_short_between:d1,d2` *DEPRECATED* does not support leap year, preferable to use `date_euro_between` or make a PR to fix it..
* `date_euro_short_max:d` *DEPRECATED* does not support leap year, preferable to use `date_euro_max` or make a PR to fix it..
* `date_euro_short_min:d` *DEPRECATED* does not support leap year, preferable to use `date_euro_min` or make a PR to fix it..
* `date_us` Date must follow the US short or long format (mm-dd-yyyy) or (mm/dd/yyyy)
* `date_us_long` Date must follow the US long format (mm-dd-yyyy) or (mm/dd/yyyy)
* `date_us_between:d1,d2` Date must follow the US short or long format and is between (d1) &amp; (d2)
* `date_us_long_between:d1,d2` Date must follow the US long format and is between (d1) &amp; (d2)
* `date_us_max:d` Date must follow US short or long format and is lower or equal than date (d)
* `date_us_long_max:d` Date must follow US long format and is lower or equal than date (d)
* `date_us_min:d` Date must follow US short or long format and is higher or equal than date (d)
* `date_us_long_min:d` Date must follow US long format and is higher or equal than date (d)
* `date_us_short` *DEPRECATED* does not support leap year, preferable to use `date_us` or make a PR to fix it.
* `date_us_short_between:d1,d2` *DEPRECATED* does not support leap year, preferable to use `date_us_between` or make a PR to fix it.
* `date_us_short_max:d` *DEPRECATED* does not support leap year, preferable to use `date_us_max` or make a PR to fix it.
* `date_us_short_min:d` *DEPRECATED* does not support leap year, preferable to use `date_us_min` or make a PR to fix it.
* `different` Alias of `different_input`
* `different_input:f` Must be different from another input field(f), where (f) must be the exact ngModel attribute of input field to compare to. The error message will use the input name or the `friendly-name` if it was provided on first input, ex.: `<input friendly-name="First Name".../>` will display :: *Field must be different from specified field "First Name"*.
* `different_input:f,t` Must be different from another input field(f), same as (different:f) but also include (t) for alternate input name to be displayed in the error message (it still uses a generic error message, if you really wish to replace the full error message then you should use `match:n:alt` see [:alt](https://github.com/ghiscoding/angular-validation/wiki/Alternate-Text-on-Validators))
* `digits:n` Ensures that field only has integer numbers and length precisely matches the specified length (n).
* `digits_between:min,max` Ensures that field only has integer numbers and is between a min,max length.
* `email` Checks for a valid email address
* `email_address` Alias of `email`
* `enum` Alias of `in_list`
* `exact_len:n` Ensures that field length precisely matches the specified length (n).
* `float` as to be floating value (excluding integer)
* `float_signed` Has to be floating value (excluding int), could be signed (-/+) positive/negative.
* ~~`iban`~~ To properly validate an IBAN please use [Wiki - Custom Validation](https://github.com/ghiscoding/angular-validation/wiki/Custom-Validation-functions) with an external library like [Github arhs/iban.js](https://github.com/arhs/iban.js)

* `in` Alias of `in_list`
* `in_list:foo,bar,..` Ensures the value is included inside the given list of values. The list must be separated by ',' and also accept words with spaces for example "ice cream".
* `int` Only positive integer (alias to `integer`).
* `integer` Only positive integer.
* `int_signed` Only integer, could be signed (-/+) positive/negative (alias to `integer_signed`).
* `integer_signed` Only integer, could be signed (-/+) positive/negative.
* `ip` Alias of `ipv4`
* `ipv4` Check for valid IP (IPv4)
* `ipv6` Check for valid IP (IPv6)
* `match:f` Match another input field(f), where (f) must be the exact ngModel attribute of input field to compare to. The error message will use the `friendly-name` if it was provided on first input, ex.: `<input friendly-name="Password".../>` will display :: *Confirmation field does not match specified field "Password"*.
* `match:f,t` Match another input field(f), same as (match:f) but also include (t) for alternate input name to be displayed in the error message (it still uses a generic error message, if you really wish to replace the full error message then you should use `match:n:alt` see [:alt](https://github.com/ghiscoding/angular-validation/wiki/Alternate-Text-on-Validators))
* `match_input` Alias of `match`.
* `max:n` Will auto-detect value type then use proper validator.
  * Type Number uses `max_num`, String use `max_len`.
* `max_date_iso` Alias of `date_iso_max`.
* `max_date_euro` Alias of `date_euro_max`.
* `max_date_euro_long` Alias of `date_euro_long_max`.
* `max_date_euro_short` *DEPRECATED* does not support leap year, preferable to use `max_date_euro` or make a PR to fix it.
* `max_date_us` Alias of `date_us_max`.
* `max_date_us_long` Alias of `date_us_long_max`.
* `max_date_us_short` *DEPRECATED* does not support leap year, preferable to use `max_date_us` or make a PR to fix it.
* `max_len:n` Checks field length, no longer than specified length where (n) is length parameter.
* `max_length:n` Alias of `max_len`
* `max_num:n` Checks numeric value to be lower or equal than the number (n).
* `min:n` Will auto-detect value type then use proper validator.
  * Type Number uses `min_num`, String use `min_len`.
* `min_date_iso` Alias of `date_iso_min`.
* `min_date_euro` Alias of `date_euro_min`.
* `min_date_euro_long` Alias of `date_euro_long_min`.
* `min_date_euro_short` *DEPRECATED* does not support leap year, preferable to use `min_date_euro` or make a PR to fix it.
* `min_date_us` Alias of `date_us_min`.
* `min_date_us_long` Alias of `date_us_long_min`.
* `min_date_us_short` *DEPRECATED* does not support leap year, preferable to use `min_date-us` or make a PR to fix it.
* `min_len:n` Checks field length, no shorter than specified length where (n) is length parameter.
* `min_length:n` Alias of `min_len`
* `min_num:n` Checks numeric value to be higher or equal than the number (n).
* `not_in` Alias of `not_in_list`

changelog.txt
Angular-Validation change logs

1.5.28 (2019-06-20) Merged PR #183, #184 fixes #185 Issue while de-registering a watcher and field level validateOnEmpty
1.5.27 (2019-04-14) Merged PR #182 to fix Radio button required field validation not working, closes #181
1.5.26 (2018-11-13) Merged PR #178 to fix revalidateAndAttachOnBlur method to handle zeroes appropriately
1.5.25 (2018-10-29) Add Chinese locale "cn.json" in the locales folder.
1.5.24 (2017-09-01) Fix #160, validate on empty not working in angular-service
1.5.23 (2017-08-17) Merged PR #162 and #163
1.5.22 (2017-06-07) Merged PR #157 to add Simplified Chinese locale.
1.5.21 (2017-05-14) Fix #151 validation rule "different" disables rule "required".
1.5.20 (2017-04-20) Add an unminified version and Fix #153 bug in validation service.
1.5.19 (2017-04-17) Fix #150 angularValidation.revalidate problems
1.5.18 (2017-03-31) Fix disabled="false" as attribute to run validation.
1.5.17 (2017-03-10) Add silent mode to checkFormValidity function
1.5.16 (2017-01-24) Fix issue where empty Form Controller was throwing an error
1.5.15 (2017-01-19) Rename errorMessageVisible flag to isErrorMessageVisible.
1.5.14 (2017-01-18) Added errorMessageVisible at updateErrorMsg this help to know when the error message is displayed or not.
1.5.13 (2016-12-29) Make sure element exist before looking for `isValidationCancelled`, issue #139
1.5.12 (2016-12-15) Fix a small issue introduced by last commit (Angular Form overwritten with simple object in some rare occasions).
1.5.11 (2016-12-15) Fix checkFormValidity with formName argument, issues #135 #139 #140 #141
1.5.10 (2016-12-13) Fix UI Router issue by adding a UI Router watch on $stateChangeStart, issue #137
1.5.9 (2016-10-14) Updated main entry at package.json to work with WebPack, issue #128
1.5.7 (2016-10-03) Add Dutch and Romanian #134
1.5.6 (2016-09-24) Dates validator now checks for leap year & full date calendar, fix #130
1.5.5 (2016-09-01) If translation isn't loading & working correctly, it should throw an error.
1.5.4 (2016-07-29) Fixed an issue with 3rd party validation not triggering a $scope.$watch. Also added new international phone validation (issue #125).
1.5.3 (2016-07-20) Add C# validation annotation, (for example: maxLen => maxLength)
1.5.2 (2016-06-14) Fixed #121 Alternate text containing the char ":" was causing unexpected displayed message.
1.5.1 (2016-03-10) Fixed #111 Add US phone number & tweaked credit card rules.
1.5.0 (2016-03-10) BREAKING CHANGE when fixing casing issue #107, ValidationService should start with uppercase. Changed to branch 1.5.x to announce major change.
1.4.22 (2016-02-02) Merged pull request #106 to add Catalan translation locale and some fixes in the Spanish locale as well.
1.4.21 (2016-01-21) Merged pull request #103 (extend Directive for revalidating) to fix request #102... thanks @tpeter1985.
1.4.20 (2016-01-17) Enhancement #100 - Add Global Option (errorMessageSeparator) for a Separator between error messages. Enhancement #101 - Add Global Option (preValidateValidationSummary) to disable pre-validation in Validation Summary if need be. Also found and fixed a problem with a try-catch throw javascript error in Custom Validation.
1.4.19 (2016-01-04) Fixed issue #99 support backslash inside `alt` (alternate message). IBAN should now be validated through custom validation (Wiki updated) with help of external library like `Github arhs/iban.js` as it was mentioned in issue [#93](https://github.com/ghiscoding/angular-validation/issues/93) (thanks @pabx06 for providing support)
1.4.18 (2015-12-20) Fixed issue #91 interpolation problem with remove validation doesn't work twice. Enhancement #98 possibility to use one validation or another (tell how many Validators are required for field to become valid). Enhancement #97 possibility to validate even on empty text field (useful on `custom` and `remote` validation). Refined some Validators (IPV6 now include compressed/uncompressed addresses, added more Polish characters to other Validators)
1.4.17 (2015-12-15) Fixed issue #92 input name with '.', enhancement #94 Polish characters, issue #96 in_list wasn't accepting special characters.
1.4.16 (2015-12-11) Fixed issue #90 blinking error messages.
1.4.15 (2015-12-02) Fixed issue #86 implicit global variable on regex.
1.4.14 (2015-11-15) Added validation-callback (#79), added #81, #82. Added new validation-callback attribute, runs after the debounce/blur and validaiton are completed. Added possibility passing arguments to Custom & Remote validators. Added new Global Options: hideErrorUnderInputs.
1.4.13 (2015-11-04) Fixed issue #78 - 'strValue is not defined' error when using `max` auto-detect validator.
1.4.12 (2015-11-01) Fixed a small issue with pulling the form name when trying to find the parent form of an input element.
1.4.11 (2015-10-29) Enhancement #75 - Added custom rules validation through custom functions. Fixed issue #76 - problem with ui-mask in directive.
1.4.10 (2015-10-12) Sanitized error messages. Fixed issue #69 - Stop when invalid characters typed on input[number].
1.4.9 (2015-10-05) Enhancement #57, #66, #67 - Added 3rd party addon validation (like ngTagsInput, Angular Multiselect, Dropdown multi-select, etc...)
1.4.8 (2015-09-12) Fixed issue #68 - Matching validation issue (password confirmation).
1.4.7 (2015-09-08) Fixed issue #65 - invalid return value on function String.format()
1.4.6 (2015-09-04) Accepted pull #64 - Update Spanish `es.json` translation file.
1.4.5 (2015-09-03) Fixed issue #63 - Custom Regex pattern should be greedy (instead of non-greedy), this was causing a problem when user add a `/` in his pattern and was stopping on first occurence.
1.4.4 (2015-08-21) Enhancement #59 - Added few Validators that are like Laravel for reusability. The `in` and `not_in` Validators are the most interesting.
1.4.3 (2015-08-18) Fixed issue #58 - multiple ControllerAs with Route change giving error of 'undefined' on $validationSummary.

full-tests/app.js
'use strict';

var myApp = angular.module('myApp', ['ngRoute', 'pascalprecht.translate', 'ghiscoding.validation']);

myApp.config(['$routeProvider', '$locationProvider', function ($routeProvider, $locationProvider) {
    $routeProvider
      .when('/validate-Directive', { templateUrl: 'Directive.html', controller: 'CtrlValidationDirective' })
      .when('/validate-Service', { templateUrl: 'Service.html', controller: 'CtrlValidationService' })
      .otherwise({ redirectTo: 'validate-Directive'  });
  }])
  .config(['$compileProvider', function ($compileProvider) {
    $compileProvider.debugInfoEnabled(false);
  }])
	.config(['$translateProvider', function ($translateProvider) {
	  $translateProvider.useStaticFilesLoader({
	    prefix: '../locales/validation/',
	    suffix: '.json'
		});

  	// load English ('en') table on startup
		$translateProvider.preferredLanguage('en');
	}])

// -- Main page Controller
// ---------------------------------------------------
myApp.controller('Ctrl', ['$location', '$route', '$scope', '$translate', function ($location, $route, $scope, $translate) {
  // change the translation language & reload the page to make sure all errors were rendered properly
  $scope.switchLanguage = function (key) {
    $translate.use(key).then(function() {
      $route.reload();
    });
  };
  $scope.goto = function ( path ) {
    $location.path( path );
  };
}]);


// -- Controller to use Angular-Validation with Directive
// -------------------------------------------------------
myApp.controller('CtrlValidationDirective', ['$scope', '$translate', 'ValidationService', function ($scope, $translate, ValidationService) {
    $scope.$validationOptions = { debounce: 50 }; // set the debounce globally with very small time for faster Protactor sendKeys()

    $scope.switchLanguage = function (key) {
      $translate.use(key);
    };

    var validatorDataset = loadData();
    $scope.validations = explodeAndFlattenValidatorArray(validatorDataset);

    $scope.submitForm = function() {
      if(new ValidationService().checkFormValidity($scope.form01)) {
        alert('All good, proceed with submit...');
      }
    }
    $scope.showValidationSummary = function () {
      $scope.displayValidationSummary = true;
    }
  }]);

// -- Controller to use Angular-Validation with Service
// -------------------------------------------------------
myApp.controller('CtrlValidationService', ['$scope', '$translate', 'ValidationService', function ($scope, $translate, ValidationService) {
    var validatorDataset = loadData();
    $scope.validations = explodeAndFlattenValidatorArray(validatorDataset);

    // start by creating the service
    var myValidation = new ValidationService();
    myValidation.setGlobalOptions({ debounce: 50, scope: $scope })

    for(var i = 0, ln = $scope.validations.length; i < ln; i++) {
      myValidation.addValidator({
        elmName: 'input' + i,
        rules: 'required|' + ($scope.validations[i].validation === 'required' ? '' : $scope.validations[i].validation)
      });
    }

    $scope.switchLanguage = function (key) {
      $translate.use(key);
    };

    $scope.submitForm = function() {
      if(new ValidationService().checkFormValidity($scope.form01)) {
        alert('All good, proceed with submit...');
      }
    }
    $scope.showValidationSummary = function () {
      $scope.displayValidationSummary = true;
    }
  }]);

// -- Private functions available to all controllers
// --------------------------------------------------

// explode the validators data, a validator might have aliases and if that is the case then exploded them into multiple validators
function explodeAndFlattenValidatorArray(data) {
  var tempArray = [];
  var obj = {};

  for(var i=0, ln = data.length; i < ln; i++) {
    obj = { validation: data[i].validator + (typeof data[i].params !== "undefined" ? ':' + data[i].params : '') };
    tempArray.push(obj);
    if(typeof data[i].aliases !== "undefined") {
      for(var j = 0, jln = data[i].aliases.length; j < jln; j++) {
        var validator = data[i].validator;
        if(!!data[i].params) {
          validator += data[i].params;
        }
        obj = { validation: data[i].aliases[j] + (typeof data[i].params !== "undefined" ? ':' + data[i].params : '') };
        tempArray.push(obj);
      }
    }
  }

  return tempArray;
}

// load dataset of all possible validators
function loadData() {
  return [
    {
      'validator': 'alpha'
    },
    {
      'validator': 'alphaSpaces',
      'aliases': ['alpha_spaces']
    },
    {
      'validator': 'alphaNum',
      'aliases': ['alpha_num']
    },
    {
      'validator': 'alphaNumSpaces',
      'aliases': ['alpha_num_spaces']
    },
    {
      'validator': 'alphaDash',
      'aliases': ['alpha_dash']
    },
    {
      'validator': 'alphaDashSpaces',
      'aliases': ['alpha_dash_spaces']
    },
    {
      'validator': 'betweenLen',
      'aliases': ['between_len'],
      'params': '1,5'
    },
    {
      'validator': 'betweenNum',
      'aliases': ['between_num'],
      'params': '5,15'
    },
    {
      'validator': 'boolean'
    },
    {
      'validator': 'creditCard',
      'aliases': ['credit_card']
    },
    {
      'validator': 'dateEuro',
      'aliases': ['date_euro']
    },
    {
      'validator': 'dateEuroBetween',
      'aliases': ['date_euro_between', 'betweenDateEuro', 'between_date_euro'],
      'params': '01-01-2000,28-02-2001'
    },
    {
      'validator': 'dateEuroMax',
      'aliases': ['date_euro_max', 'maxDateEuro', 'max_date_euro'],
      'params': '30-05-2012'
    },
    {
      'validator': 'dateEuroMin',
      'aliases': ['date_euro_min', 'minDateEuro', 'min_date_euro'],
      'params': '25-05-2012'
    },
    {
      'validator': 'dateEuroShort',
      'aliases': ['date_euro_short'],
      'params': '25-05-12'
    },
    {
      'validator': 'dateEuroShortBetween',
      'aliases': ['date_euro_short_between', 'betweenDateEuroShort', 'between_date_euro_short'],
      'params': '25-05-12,04-06-12'
    },
    {
      'validator': 'dateEuroShortMax',
      'aliases': ['date_euro_short_max', 'maxDateEuroShort', 'max_date_euro_short'],
      'params': '04-06-12'
    },
    {
      'validator': 'dateEuroShortMin',
      'aliases': ['date_euro_short_min', 'minDateEuroShort', 'min_date_euro_short'],
      'params': '04-06-12'
    },
    {
      'validator': 'dateIso',
      'aliases': ['date_iso']
    },
    {
      'validator': 'dateIsoBetween',
      'aliases': ['date_iso_between', 'betweenDateIso', 'between_date_iso'],
      'params': '2012-05-25,2012-06-04'
    },
    {
      'validator': 'dateIsoMax',
      'aliases': ['date_iso_max', 'maxDateIso', 'max_date_iso'],
      'params': '2012-05-25'
    },
    {
      'validator': 'dateIsoMin',
      'aliases': ['date_iso_min', 'minDateIso', 'min_date_iso'],
      'params': '2012-05-25'
    },
    {
      'validator': 'dateUs',
      'aliases': ['date_us']
    },
    {
      'validator': 'dateUsBetween',
      'aliases': ['date_us_between', 'betweenDateUs', 'between_date_us'],
      'params': '01/01/1990,12/31/2015'
    },
    {
      'validator': 'dateUsMax',
      'aliases': ['date_us_max', 'maxDateUs', 'max_date_us'],
      'params': '01/01/1990'
    },
    {
      'validator': 'dateUsMin',
      'aliases': ['date_us_min', 'minDateUs', 'min_date_us'],
      'params': '01/01/1990'
    },
    {
      'validator': 'dateUsShort',
      'aliases': ['date_us_short']
    },
    {
      'validator': 'dateUsShortBetween',
      'aliases': ['date_us_short_between', 'betweenDateUsShort', 'between_date_us_short'],
      'params': '01/01/90,12/31/15'
    },
    {
      'validator': 'dateUsShortMax',
      'aliases': ['date_us_short_max', 'maxDateUsShort', 'max_date_us_short'],
      'params': '01/01/90'
    },
    {
      'validator': 'dateUsShortMin',
      'aliases': ['date_us_short_min', 'minDateUsShort', 'min_date_us_short'],
      'params': '01/01/90'
    },
    {
        'validator': 'digits',
        'params': '3'
      },
      {
        'validator': 'digitsBetween',
        'aliases': ['digits_between'],
        'params': '1,5'
      },
    {
      'validator': 'email'
    },
    {
      'validator': 'exactLen',
      'aliases': ['exact_len'],
      'params': '11'
    },
    {
      'validator': 'float'
    },
    {
      'validator': 'floatSigned',
      'aliases': ['float_signed']
    },
    {
      'validator': 'iban'
    },
    {
      'validator': 'in',
      'aliases': ['inList', 'in_list'],
      'params': 'chocolate,apple pie,ice cream,sweet & sour,A+B'
    },
    {
      'validator': 'int',
      'aliases': ['integer']
    },
    {
      'validator': 'intSigned',
      'aliases': ['integerSigned', 'int_signed', 'integer_signed']
    },
    {
      'validator': 'ipv4',
      'aliases': ['ip']
    },
    {
      'validator': 'ipv6'
    },
    {
      'validator': 'maxLen',
      'aliases': ['max_len'],
      'params': '11'
    },
    {
      'validator': 'maxNum',
      'aliases': ['max_num'],
      'params': '99'
    },
    {
      'validator': 'minLen',
      'aliases': ['min_len'],
      'params': '3'
    },
    {
      'validator': 'minNum',
      'aliases': ['min_num'],
      'params': '1'
    },
    {
      'validator': 'notIn',
      'aliases': ['not_in', 'notInList', 'not_in_list'],
      'params': 'chocolate,apple pie,ice cream,sweet & sour,A+B'
    },
    {
      'validator': 'numeric'
    },
    {
      'validator': 'numericSigned',
      'aliases': ['numeric_signed']
    },
    {
      'validator': 'phone'
    },
    {
      'validator': 'url'
    },
    {
      'validator': 'time'
    }
  ];
}